### PR TITLE
feat/ibc-hooks #ntrn-344

### PR DESCRIPTION
**Pull request by NeverHappened**
Implements [this task](https://p2pvalidator.atlassian.net/browse/NTRN-344)

Adds ibc-hooks module, code was copied from osmosis (see module README.md for details on commit and version).

Basically we add middleware that accepts incoming ics20 packets and if they have a proper memo with encoded cosmwasm message, call the contract.

Test run:
https://github.com/neutron-org/neutron-tests/actions/runs/4606069323

### CODE REVIEW

File: x/ibc-hooks/wasm_hook.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+	if err != nil {
+		// The tokens will be returned
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, err.Error())
+	}
+
+	return isWasmRouted, contractAddr, msgBytes, nil
+}
+
+func (h WasmHooks) SendPacketOverride(i ICS4Middleware, ctx sdk.Context, chanCap *capabilitytypes.Capability, packet ibcexported.PacketI) error {
```
Comment (NeverHappened) [inline x/ibc-hooks/wasm_hook.go]: @oopcode @pr0n00gler 
This implements this ibc_callback feature. Do we need it?
It works like this: we send message with memo: {"ibc_callback": "ntrncontractaddress"},
and will get a sudo callback on this address later.
We already kinda have same functionality, right?

Comment (mmikev) [reply x/ibc-hooks/wasm_hook.go]: Let's disable callbacks from hooks on Neutron side.
So if there is a transfer from Neutron to some other chain, a callback will happen through our system.

File: x/ibc-hooks/README.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 Taken from [osmosis](https://github.com/osmosis-labs/osmosis) `v14.0.0-rc1` (commit `26e2fad8e7b3eb7c33965360b31a593b392d7d75`)
 
-Removed `ibc_callback` functionality
+Removed `ibc_callback` functionality.
```
Comment (mmikev) [inline x/ibc-hooks/README.md]: Please, leave some info why removed that

**Review by foxpy — DISMISSED** [review]

**Review by ratik — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

File: x/ibc-hooks/sdkmodule.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+var (
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
+)
+
+// AppModuleBasic defines the basic application module used by the ibc-hooks module.
+type AppModuleBasic struct{}
+
+var _ module.AppModuleBasic = AppModuleBasic{}
```
Comment (swelf19) [inline x/ibc-hooks/sdkmodule.go:31]: var declared second time

File: x/ibc-hooks/sdkmodule.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	return cli.GetQueryCmd()
+}
+
+// ___________________________________________________________________________
+
+// AppModule implements an application module for the ibc-hooks module.
+type AppModule struct {
+	AppModuleBasic
+
+	authKeeper types.AccountKeeper
```
Comment (swelf19) [inline x/ibc-hooks/sdkmodule.go:76]: why do you need this field, 
`types.AccountKeeper` interface
and the wrapper module at all

File: x/ibc-hooks/hooks.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
+	// ibc-go
+	channeltypes "github.com/cosmos/ibc-go/v4/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v4/modules/core/exported"
+)
+
+type Hooks interface{}
+
+type OnChanOpenInitOverrideHooks interface {
```
Comment (swelf19) [inline x/ibc-hooks/hooks.go:15]: almost all interfaces are not implemented

**Review by swelf19 — APPROVED** [review]

**Pull request merged**